NanoProd/plugins/TauSpinnerTableProducer.cc

```diff
@@ -0,0 +1,245 @@
+/**\class TauSpinnerTableProducer
+
+ Description: Produces FlatTable with TauSpinner weights for H->tau,tau events
+
+ Original Author: D.  Winterbottom (IC)
+                  update, M. Bluj (NCBJ)
+*/
+
+#include <memory>
+#include <vector>
+#include <string>
+
+#include "boost/functional/hash.hpp"
+#include "boost/format.hpp"
+
+#include "FWCore/Framework/interface/one/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Concurrency/interface/SharedResourceNames.h"
+#include "DataFormats/Common/interface/View.h"
+#include "DataFormats/HepMCCandidate/interface/GenParticle.h"
+#include "DataFormats/NanoAOD/interface/FlatTable.h"
+
+#include "Tauola/Tauola.h"
+#include "TauSpinner/SimpleParticle.h"
+#include "TauSpinner/tau_reweight_lib.h"
+
+namespace {
+  std::set<int> tauProdsIds{22, 111, 211, 321, 130, 310, 11, 12, 13, 14, 16};
```

Comment:
```suggestion
  const std::set<int> tauProdsIds{22, 111, 211, 321, 130, 310, 311, 321, 11, 12, 13, 14, 16};
```
Added `K0` and `K` that can occur in tau decays. Based on TauMLTools code (which was run on the full DeepTau training statistics) the list should be complete now.

Reply:
Interestingly, I have run a small test (100 HTT events) and eta (pdg=221) was found among tau decay products... Accordingly to PDG most probable decay with eta (eta pi+- pi0) has BR of ~1.4e-3, so it can appear in simulation. So, I am adding it to the list.

Reply:
Very interesting! Thanks for the cross-check. Indeed, after looking back in the code, the pdg-id check was removed from TauMLTools implementation at some point, so it was not tested on the whole training dataset... I ran a quick check on HTT and DY samples and found that in addition to `eta` (221), we need to add `omega` (223) and `K_star` (323).

Reply:
OK, thanks I will add them to the list. However, I wonder what is a "correct" list. In the original implementation at least some of these particles were catched via their decay products (pions, Kaons, photons), i.e. the were treated as intermediate resonances. I don't know what tauola/tauspinner expects...

Reply:
I checked the [TauSpinner code](https://tauolapp.web.cern.ch/tau__reweight__lib_8cxx_source.html#l00814), and indeed it expects only kaons and pions. However, K0 (311) is missing in the original set, while it is supported by the TauSpinner. I suggest that we identify supported decays at the stage of producer, and if the decay of at least one of two taus is not supported (e.g. direct `eta->gamma gamma` decays), we can agree to store some default value to be able to identify such events at the analysis level). I propose to modify `getTauDaughters` as following:
```c++
bool TauSpinnerTableProducer::getTauDaughters(reco::GenParticleRefVector &tau_daughters, const reco::GenParticle &part) {
  static const std::set<int> directTauProducts = { 11, 12, 13, 14, 16, 22 };
  static const std::set<int> finalHadrons = { 111, 130, 211, 310, 311, 321 };
  static const std::set<int> intermediateHadrons = { 221, 223, 323 };

  for (auto daughterRef : part.daughterRefVector()) {
    const int daughter_pdgId = std::abs(daughterRef->pdgId());
    if( (std::abs(part.pdgId()) == 15 && directTauProducts.count(daughter_pdgId))
        || finalHadrons.count(daughter_pdgId)) {
      tau_daughters.push_back(daughterRef);
    } else if (intermediateHadrons.count(daughter_pdgId)) {
      if(!getTauDaughters(tau_daughters, *daughter))
        return false;
    } else {
      // can printout something in LogWarning
      return false;
    }
  }
  return true;
}
```

Reply:
Thanks for checking. I tried to found it in documentation, but I failed. 
We can proceed as you propose - do you have a special weight value to be used, 0 (zero)? It will simply remove problematic events when not checked and I think that it will have a negligible effect as I guess that no-supported decays should be quite rare.

Reply:
Zero as the default value is fine for me. If needed, this logic can be repeated at the nanoAOD level (since the full tau decay chain is stored) to verify the origin of 0.

Reply:
I run test with ~660 HTT events and there are two events with one tau decaying with eta->gamma,gamma in the chain. It is compatible with BR(tau- ->pi- eta pi0) of ~1.8e-3 and BR(eta->gamma,gamma) ~0.4. I wonder if such taus are efficiently reconstructed (I suppose that not) and to which DM they enter.

Reply:
Such investigation would be indeed interesting, but I think it is beyond the scope of this PR.

Reply:
Yes, sure.
Anyway, I will inform people about implementation choice made at the CP HTT mattermost chat.